tango/integrations/transformers/ia3.py

```diff
@@ -0,0 +1,244 @@
+import re
+from dataclasses import dataclass
+from typing import Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from transformers import PreTrainedModel
+from transformers.modeling_utils import Conv1D
+
+
+@dataclass
+class WithIA3Config:
+    ia3_param_names: str
+    attention_modules: str
+    mlp_modules: str
+    mlp_layers: str
+    fused_qkv_layers: Optional[str] = None
+    k_layers: Optional[str] = None
+    v_layers: Optional[str] = None
+
+
+GPT_J_IA3_CONFIG = WithIA3Config(
+    attention_modules=".*attn",
+    k_layers="k_proj",
+    v_layers="v_proj",
+    mlp_modules=".*mlp",
+    mlp_layers="fc_in",
+    ia3_param_names="ia3",
+)
+
+GPT_2_IA3_CONFIG = WithIA3Config(
+    attention_modules=".*attn",
+    fused_qkv_layers="c_attn",
+    mlp_modules=".*mlp",
+    mlp_layers="c_fc",
+    ia3_param_names="ia3",
+)
+
+OPT_IA3_CONFIG = WithIA3Config(
+    attention_modules=".*self_attn",
+    k_layers="k_proj",
+    v_layers="v_proj",
+    mlp_modules=r".*layers\.\d*",
+    mlp_layers="fc1",
+    ia3_param_names="ia3",
+)
+
+BLOOM_IA3_CONFIG = WithIA3Config(
+    attention_modules=".*self_attention",
+    fused_qkv_layers="query_key_value",
+    mlp_modules=".*mlp",
+    mlp_layers="dense_h_to_4h",
+    ia3_param_names="ia3",
+)
+
+MODEL_NAME_TO_CONFIG = {
+    "sshleifer/tiny-gpt2": GPT_2_IA3_CONFIG,
+    "gpt2": GPT_2_IA3_CONFIG,
+    "gpt2-medium": GPT_2_IA3_CONFIG,
+    "gpt2-large": GPT_2_IA3_CONFIG,
+    "gpt2-xl": GPT_2_IA3_CONFIG,
+    "bigscience/bloom-560m": BLOOM_IA3_CONFIG,
+    "bigscience/bloom-1b1": BLOOM_IA3_CONFIG,
+    "bigscience/bloom-1b7": BLOOM_IA3_CONFIG,
+    "bigscience/bloom-3b": BLOOM_IA3_CONFIG,
+    "bigscience/bloom-7b1": BLOOM_IA3_CONFIG,
+    "bigscience/bloom": BLOOM_IA3_CONFIG,
+    "facebook/opt-125m": OPT_IA3_CONFIG,
+    "facebook/opt-350m": OPT_IA3_CONFIG,
+    "facebook/opt-1.3b": OPT_IA3_CONFIG,
+    "facebook/opt-2.7b": OPT_IA3_CONFIG,
+    "facebook/opt-6.7b": OPT_IA3_CONFIG,
+    "facebook/opt-13b": OPT_IA3_CONFIG,
+    "facebook/opt-30b": OPT_IA3_CONFIG,
+    "facebook/opt-66b": OPT_IA3_CONFIG,
+    "EleutherAI/gpt-j-6B": GPT_J_IA3_CONFIG,
+}
+
+
+class LinearWithIA3(nn.Module):
+    def __init__(self, linear_layer, ia3_param_names, unfuse_size: int = None):
```

Comment:
Docstring should be added. It took me a moment to understand that `unfuse_size` was the `hidden_size`.